### PR TITLE
update shared convolution

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?

update shared convolution with latest SpatialConvolution.
Add wRegularize, bRegularizer, initWeight, initBias, initGradWeight, initGradBias, withBias

## How was this patch tested?

unit tests

### CODE REVIEW

**Comment by jason-dai** [comment]
Please add more more descriptions on what is exactly changed in the PR.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 1]
```diff
@@ -16,112 +16,87 @@
...
+      require(outputWidth >= 1 && outputHeight >= 1,
+        s"output size is too small. outputWidth: $outputWidth, outputHeight: $outputHeight")
+
+      if (onesBias.dim() != 1 || onesBias.size(1) != outputHeight * outputWidth) {
+        onesBias.resize(Array(outputHeight * outputWidth)).fill(ev.fromType(1.0))
       }
-    } else {
       require(input.size(2) == nInputPlane)
       val batchSize = input.size(1)
       output.resize(Array(batchSize, nOutputPlane, outputHeight, outputWidth))
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: ` output.resize(batchSize, nOutputPlane, outputHeight, outputWidth)`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 2]
```diff
@@ -16,112 +16,87 @@
...
-
-      val coresNum = Math.min(batchSize, Engine.model.getPoolSize)
-      fInput.resize(Array(coresNum, nGroup, kernelW * kernelH * nInputPlane / nGroup,
+      fInput.resize(Array(nGroup, kernelW * kernelH * nInputPlane / nGroup,
         outputHeight * outputWidth))
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: fInput.resize(nGroup, kernelW * kernelH * nInputPlane / nGroup,
          outputHeight * outputWidth)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 3]
```diff
@@ -130,106 +105,47 @@ class SpatialShareConvolution[T: ClassTag](
...
-          ev.fromType[Double](scaleB))
-        g += 1
-      }
+    if (_1x1 || input.dim() == 3 || (input.dim() == 4 && input.size(1) == 1)) {
+      super.accGradParameters(input, gradOutput)
     } else {
       val batchSize = input.size(1)
       if (gradWeightMMInBatch == null) {
         gradWeightMMInBatch = Tensor[T]().resize(Array(batchSize, nGroup, nOutputPlane / nGroup,
           nInputPlane * kernelH * kernelW / nGroup))
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]:
```
gradWeightMMInBatch = Tensor[T]().resize(batchSize, nGroup, nOutputPlane / nGroup,
            nInputPlane * kernelH * kernelW / nGroup)
```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 4]
```diff
@@ -130,106 +105,47 @@ class SpatialShareConvolution[T: ClassTag](
...
       }
-      if (gradientBiasMT.nElement() == 0) {
+      if(gradientBiasMT.nElement() == 0) {
         gradientBiasMT.resize(Array(batchSize, nOutputPlane))
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: `gradientBiasMT.resize(batchSize, nOutputPlane)`

**Comment by yangw1234** [comment]
also changing python api?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 5]
```diff
@@ -239,72 +155,77 @@ class SpatialShareConvolution[T: ClassTag](
...
-                ev.fromType[Double](scaleW),
-                ev.fromType[Double](scaleB))
-              g += 1
-            }
-            _i += 1
+      var i = 1
+      while (i <= batchSize) {
+        val inputT = input.select(1, i)
+          val gradOutputT = gradOutput.select(1, i)
+          var g = 0
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: `val gradWeightMMInBatchT = gradWeightMMInBatch.select(1, i)`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 6]
```diff
@@ -239,72 +155,77 @@ class SpatialShareConvolution[T: ClassTag](
...
+              fInput.select(1, g + 1),
+              kernelW, kernelH, strideW, strideH,
+              padW, padH,
+              nInputPlane / nGroup, inputWidth, inputHeight,
+              nOutputPlane / nGroup, outputWidth, outputHeight)
+
+            calcGradParametersFrame(
+              gradOutputT.narrow(1, g * nOutputPlane / nGroup + 1, nOutputPlane / nGroup),
+              gradWeightMMInBatch.select(1, i).select(1, g + 1),
+              gradientBiasMT.select(1, i).narrow(1, g * nOutputPlane / nGroup + 1,
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]:
```
gradWeightMMInBatchT.select(1, g + 1)
gradientBiasMTT.narrow(1, g * nOutputPlane / nGroup + 1,nOutputPlane / nGroup)
```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 7]
```diff
@@ -391,10 +316,74 @@ object SpatialShareConvolution {
...
+      conv.strideW, conv.strideH,
+      conv.padW, conv.padH,
+      conv.nGroup, conv.propagateBack,
+      conv.wRegularizer, conv.bRegularizer
+    )
+    sConv.weight.copy(conv.weight)
+    sConv.bias.copy(conv.bias)
+    sConv.setScaleW(conv.getScaleW())
+    sConv.setScaleB(conv.getScaleB())
+    sConv
```
Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: setName

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 8]
```diff
@@ -130,210 +109,102 @@ class SpatialShareConvolution[T: ClassTag](
...
       val batchSize = input.size(1)
       if (gradWeightMMInBatch == null) {
-        gradWeightMMInBatch = Tensor[T]().resize(Array(batchSize, nGroup, nOutputPlane / nGroup,
-          nInputPlane * kernelH * kernelW / nGroup))
+        gradWeightMMInBatch = Tensor[T](batchSize, nGroup, nOutputPlane / nGroup,
+          nInputPlane * kernelH * kernelW / nGroup)
       }
-      if (gradientBiasMT.nElement() == 0) {
-        gradientBiasMT.resize(Array(batchSize, nOutputPlane))
+      if(withBias && gradientBiasMT.nElement() == 0) {
```
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: add a space after if ?
`if (withBias && gradientBiasMT.nElement() == 0)`

**Comment by qiuxin2012** [comment]
jenkins passed

**Pull request merged**